### PR TITLE
[Merged by Bors] - feat : port Algebra.GroupPower.Basic

### PR DESCRIPTION
**Pull request by javra**
mathlib3 SHA: 9b2660e1b25419042c8da10bf411aa3c67f14383

### CODE REVIEW

**Comment by javra** [comment]
Removed some `@[simp]` which violated SimpNF (and as far as I can see did so in mathlib3?)

**Comment by javra** [comment]
(couldn't commit the raw `mathlib3port` file due to the file already being half ported)

File: Mathlib/Algebra/GroupPower/Basic.lean [hunk 1]
```diff
@@ -3,71 +3,454 @@ Copyright (c) 2015 Jeremy Avigad. All rights reserved.
...
   · rw [Nat.add_zero, pow_zero, mul_one]
   · rw [pow_succ', ← mul_assoc, ← ih, ← pow_succ', Nat.add_assoc]
 
+@[to_additive mul_nsmul']
 theorem pow_mul (a : M) (m n : ℕ) : a ^ (m * n) = (a ^ m) ^ n := by
   induction' n with n ih
   · rw [Nat.mul_zero, pow_zero, pow_zero]
   · rw [Nat.mul_succ, pow_add, pow_succ', ih]
 
+@[to_additive mul_nsmul]
```
Comment (ChrisHughes24) [inline Mathlib/Algebra/GroupPower/Basic.lean:117]: Should this be `@[to_additive mul_smul']` and the attribute above should also be changed?

File: Mathlib/Algebra/GroupPower/Basic.lean [hunk 2]
```diff
@@ -3,71 +3,454 @@ Copyright (c) 2015 Jeremy Avigad. All rights reserved.
...
+#align of_add_zsmul of_add_zsmul
+
+theorem of_mul_pow [Monoid A] (x : A) (n : ℕ) : Additive.ofMul (x ^ n) = n • Additive.ofMul x :=
+  rfl
+#align of_mul_pow of_mul_pow
+
+theorem of_mul_zpow [DivInvMonoid G] (x : G) (n : ℤ) :
+    Additive.ofMul (x ^ n) = n • Additive.ofMul x :=
+  rfl
+#align of_mul_zpow of_mul_zpow
```
Comment (ChrisHughes24) [inline Mathlib/Algebra/GroupPower/Basic.lean:409]:
```suggestion
theorem ofAdd_nsmul [AddMonoid A] (x : A) (n : ℕ) :
    Multiplicative.ofAdd (n • x) = Multiplicative.ofAdd x ^ n :=
  rfl
#align of_add_nsmul ofAdd_nsmul

theorem ofAdd_zsmul [SubNegMonoid A] (x : A) (n : ℤ) :
    Multiplicative.ofAdd (n • x) = Multiplicative.ofAdd x ^ n :=
  rfl
#align of_add_zsmul ofAdd_zsmul

theorem ofMul_pow [Monoid A] (x : A) (n : ℕ) : Additive.ofMul (x ^ n) = n • Additive.ofMul x :=
  rfl
#align of_mul_pow ofMul_pow

theorem ofMul_zpow [DivInvMonoid G] (x : G) (n : ℤ) :
    Additive.ofMul (x ^ n) = n • Additive.ofMul x :=
  rfl
#align of_mul_zpow ofMul_zpow
```

File: Mathlib/Algebra/GroupPower/Basic.lean (hunk 1 above)
Comment (semorrison) [reply Mathlib/Algebra/GroupPower/Basic.lean]: I did this, and left a porting note.

File: Mathlib/Algebra/GroupPower/Basic.lean [hunk 3]
```diff
@@ -3,71 +3,500 @@ Copyright (c) 2015 Jeremy Avigad. All rights reserved.
...
+
+universe u v w x y z u₁ u₂
+
+variable {α : Type _} {M : Type u} {N : Type v} {G : Type w} {H : Type x} {A : Type y} {B : Type z}
+  {R : Type u₁} {S : Type u₂}
+
+/-!
+### Commutativity
+
+First we prove some facts about `semiconj_by` and `commute`. They do not require any theory about
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/GroupPower/Basic.lean:40]:
```suggestion
First we prove some facts about `SemiconjBy` and `Commute`. They do not require any theory about
```

File: Mathlib/Algebra/GroupPower/Basic.lean [hunk 4]
```diff
@@ -3,71 +3,500 @@ Copyright (c) 2015 Jeremy Avigad. All rights reserved.
...
+# Power operations on monoids and groups
+
+The power operation on monoids and groups.
+We separate this from group, because it depends on `ℕ`,
+which in turn depends on other parts of algebra.
+
+This module contains lemmas about `a ^ n` and `n • a`, where `n : ℕ` or `n : ℤ`.
+Further lemmas can be found in `algebra.group_power.lemmas`.
+
+The analogous results for groups with zero can be found in `algebra.group_with_zero.power`.
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/GroupPower/Basic.lean:20]:
```suggestion
The analogous results for groups with zero can be found in `Algebra.GroupWithZero.Power`.
```

File: Mathlib/Algebra/GroupPower/Basic.lean [hunk 5]
```diff
@@ -3,71 +3,500 @@ Copyright (c) 2015 Jeremy Avigad. All rights reserved.
...
+
+## Notation
+
+- `a ^ n` is used as notation for `has_pow.pow a n`; in this file `n : ℕ` or `n : ℤ`.
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/GroupPower/Basic.lean:24]:
```suggestion
- `a ^ n` is used as notation for `Pow.pow a n`; in this file `n : ℕ` or `n : ℤ`.
```

File: Mathlib/Algebra/GroupPower/Basic.lean [hunk 6]
```diff
@@ -3,71 +3,500 @@ Copyright (c) 2015 Jeremy Avigad. All rights reserved.
...
+- `n • a` is used as notation for `has_smul.smul n a`; in this file `n : ℕ` or `n : ℤ`.
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/GroupPower/Basic.lean:25]:
```suggestion
- `n • a` is used as notation for `SMul.smul n a`; in this file `n : ℕ` or `n : ℤ`.
```

File: Mathlib/Tactic/Abel.lean [hunk 7]
```diff
@@ -251,7 +251,6 @@ theorem term_atom {α} [AddCommMonoid α] (x : α) : x = term 1 x 0 := by
 theorem term_atomg {α} [AddCommGroup α] (x : α) : x = termg 1 x 0 := by
   -- TODO waiting for port of Algebra.GroupPower.Basic for `one_zsmul`
```
Comment (Ruben-VandeVelde) [inline Mathlib/Tactic/Abel.lean:252]:
```suggestion
```

**Review by Ruben-VandeVelde — APPROVED** [review]

**Comment by javra** [comment]
Thanks for those, hadn't touched the comments and docstsrings yet.

**Comment by semorrison** [comment]
> > There are lots of additional #aligns needed for the @[to_additive] versions. Hopefully this will be automated in mathport later: https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/.23align.20and.20to_additive. But for now we just need to do it by hand. :-(
> 
> Thanks! I didn't really get the memo that we have to add _all_ aligns generally instead of just the ones that deviate from the synport-generated ones.

We don't really need to add all of them. It seems we've just decided it's much easier to review / help out others PRs / maintain later, if all #align's are there.

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**